msdm/algorithms/policyiteration.py

```diff
@@ -1,96 +1,139 @@
-            if self.check_unreachable_convergence:
-                converged = (new_pi == pi).all()
-            else:
-                converged = (new_pi[rs, :] == pi[rs, :]).all()
-            if converged:
-                break
-            pi = new_pi
+        results = []
+        for mdp, state_values, action_values, policy_matrix in zip(
+            mdps, state_values_batch, action_values_batch, policy_matrix_batch
+        ):
+            state_values[mdp.recurrent_state_vec,] = self.undefined_value
+            action_values[mdp.recurrent_state_vec,] = self.undefined_value
+            policy=TabularPolicy.from_state_action_lists(
+                state_list=mdp.state_list,
+                action_list=mdp.action_list,
+                data=policy_matrix
+            )
+            state_values=StateTable.from_state_list(
+                state_list=mdp.state_list,
+                data=state_values
+            )
+            action_values=StateActionTable.from_state_action_lists(
+                state_list=mdp.state_list,
+                action_list=mdp.action_list,
+                data=action_values
+            )
+            results.append(PolicyIterationResult(
+                iterations=iterations,
+                state_value=state_values,
+                action_value=action_values,
+                converged=iterations < (self.max_iterations - 1),
+                initial_value=sum([state_values[s]*p for s, p in mdp.initial_state_dist().items()]),
+                policy=policy
+            ))
+        return results
 
-        validq = q + np.log(am)
-        pi = TabularPolicy.from_q_matrix(mdp.state_list, mdp.action_list, validq)
+@dataclass
+class PolicyIterationResult(PlanningResult):
+    iterations : int
+    converged : bool
+    state_value : dict
+    initial_value : float
+    action_value: dict
+    policy : TabularPolicy
 
-        # create result object
-        res = PlanningResult()
-        if i == (iterations - 1):
-            warnings.warn(f"Policy Iteration not converged after {iterations} iterations")
-            res.converged = False
-        else:
-            res.converged = True
-        res.mdp = mdp
-        res.policy = res.pi = pi
-        res._valuevec = v
-        vf = dict()
-        for s, vi in zip(mdp.state_list, v):
-            vf[s] = vi
-        res.valuefunc = res.V = vf
-        res._qvaluemat = q
-        res.iterations = i
-        qf = defaultdict(lambda : dict())
-        for si, s in enumerate(mdp.state_list):
-            for ai, a in enumerate(mdp.action_list):
-                qf[s][a] = q[si, ai]
-        res.actionvaluefunc = res.Q = qf
-        res.initial_value = sum([res.V[s0]*p for s0, p in mdp.initial_state_dist().items()])
-        return res
+def policy_iteration_vectorized(
+    transition_matrix,
+    discount_rate,
+    state_action_reward_matrix,
+    action_matrix,
+    policy_matrix,
+    max_iterations=int(1e5),
+    value_difference=1e-10,
```

Comment:
I think standardizing across algorithms, to the extent that that makes any sense? Like, why is this `1e-10` instead of the `1e-8` that's `numpy`'s default? So maybe standardizing + a comment explaining it?

Reply:
See 05a8b4a - I decided to get rid of `VALUE_DECIMAL_PRECISION` in both `PolicyIteration` and `ValueIteration`, and just use the numpy `isclose` defaults, since it's use was for cases where numerical stability was important (e.g., comparing equality to the max value). Keeping `max_residual` in `ValueIteration` since that's part of the definition of the algorithm/convergence.